File: Sources/Features/Projects/UseCases/CreateProjectUseCase.swift
// MARK: - CreateProjectUseCase
// Use case for creating a new project directory and registering it.
// macOS 14+, Swift 5.10

import Foundation

/// Creates a new project directory on disk and registers it in the project manager.
///
/// Extracted from `CreateNewProjectSheet` to keep the view free of filesystem
/// operations — all business logic lives in the use case layer.
@MainActor
struct CreateProjectUseCase {

    let projectManager: any ProjectManaging

    /// Create a new project directory at `parent/name` and activate it.
    ///
    /// - Parameters:
    ///   - name: The folder name (already trimmed, non-empty).
    ///   - parent: Parent directory where the folder will be created.
    /// - Throws: `FileManager` error if creation fails, or `ProjectManagerError` on duplicate.
    @discardableResult
    func execute(name: String, parent: URL) throws -> Project {
        let newURL = parent.appendingPathComponent(name)
        try FileManager.default.createDirectory(
            at: newURL,
            withIntermediateDirectories: true,
            attributes: nil
        )
        let project = try projectManager.addProject(at: newURL)
        projectManager.activeProjectId = project.id
        return project
    }
}

File: shared/src/commonMain/kotlin/studio/vibe/shared/usecase/CreateProjectUseCase.kt
package studio.vibe.shared.usecase

import studio.vibe.shared.contract.PersistenceStore
import studio.vibe.shared.contract.ProjectManaging
import studio.vibe.shared.model.FilePath
import studio.vibe.shared.model.Project
import kotlin.uuid.ExperimentalUuidApi

/**
 * Creates a new project directory on disk and registers it in the project manager.
 *
 * Extracted from the create-project UI sheet to keep views free of filesystem
 * operations — all business logic lives in the use case layer.
 */
@OptIn(ExperimentalUuidApi::class)
class CreateProjectUseCase(
    private val projectManager: ProjectManaging,
    private val persistence: PersistenceStore,
) {
    /**
     * Create a new project directory at `parentPath/name` and activate it.
     *
     * @param name The folder name (already trimmed, non-empty).
     * @param parentPath Parent directory where the folder will be created.
     * @return The newly created and activated [Project].
     * @throws Exception if directory creation fails or the project already exists.
     */
    suspend fun execute(name: String, parentPath: FilePath): Project {
        val newPath = parentPath.child(name)
        persistence.createDirectory(newPath)
        val project = projectManager.addProject(newPath)
        projectManager.setActiveProjectId(project.id)
        return project
    }
}
